### PR TITLE
[OpenMP][MLIR] Descriptor explicit member map lowering changes

### PR DESCRIPTION
**Pull request by agozillon**
This is one of 3 PRs in a PR stack that aims to add support for explicit mapping of
allocatable members in derived types.

The primary changes in this PR are the OpenMPToLLVMIRTranslation.cpp changes,
which are small and seek to alter the current member mapping to add an
additional map insertion for pointers. Effectively, if the member is a pointer
(currently indicated by having a varPtrPtr field) we add an additional map for
the pointer and then alter the subsequent mapping of the member (the data)
to utilise the member rather than the parents base pointer. This appears to be
necessary in certain cases when mapping pointer data within record types to
avoid segfaulting on device (due to incorrect data mapping). In general this
record type mapping may be simplifiable in the future.

There are also additions of tests which should help to showcase the affect
of the changes above.

### CODE REVIEW

File: mlir/test/Target/LLVMIR/omptarget-fortran-allocatable-record-type-mapping-host.mlir
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/omptarget-fortran-allocatable-record-type-mapping-host.mlir:1]: Please reduce this test to the absolute minimum that still tests your change. The IR does not need to compute anything useful, but the test should fail when you would revert only your C++ changes.

Additionally, try to reduce the `CHECK`s to a  minimum to not make the test exceptionally sensitive to all kinds of changes.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2373]: Nit: It would be nice to add a short comment of what's achieved by this block of code.

**Review by skatrak — APPROVED** [review]
Thank you Andrew, this looks reasonable to me, though I'd suggest waiting for another approval before merging.

Also, it looks like the flang/test/Integration/OpenMP/map-types-and-sizes.f90 test needs to be updated.

**Review by TIFitis — APPROVED** [review]
LGTM with the other comments addressed :)

**Comment by agozillon** [comment]
> Thank you Andrew, this looks reasonable to me, though I'd suggest waiting for another approval before merging.
> 
> Also, it looks like the flang/test/Integration/OpenMP/map-types-and-sizes.f90 test needs to be updated.

Thank you, I'll add the comment and simplify the test as previously requested by another reviewer :-) However, the test doesn't need to be updated I believe, it's just a side affect of this being the middle PR stack! That test unfortunately is sensitive to both MLIR -> LLVM-IR changes and Flang -> MLIR changes as it tests the full lowering process! And as it's in Flang's test infrastructure it's in the Flang portion of the changes, you should see it pass (alongside all other tests) in: https://github.com/llvm/llvm-project/pull/96266/ which is built on top of this via SPR!

**Comment by agozillon** [comment]
Thank you for the approvals and comments! I updated the PR to add the requested comment as well as shortened the test as requested to a more minimal test.

**Comment by agozillon** [comment]
Likely also worth stating the PR has some further modifications to support swapping from DenseIntElementsAttr -> ArrayAttr<ArrayAttr<IntegerAttr>>> which simplifies the implementation a chunk by no longer requiring the -1 padding (thank you very much for the suggestions @skatrak :-) )! So perhaps worth another skim over the PR.

**Comment by agozillon** [comment]
Small ping for some reviewer attention on this PR please, if at all possible. Just to make sure the current PR approvals stand after recent modifications. Thank you ahead of time as always! :-)

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2282]: Nit: I think `llvm::zip` could simplify this a bit, since it already implements iterating over two ranges of potentially different sizes until the end of the shortest one is reached.

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2548]:
```suggestion
    // subsequent map to the pointer. This segment of code generates the
    // pointer mapping, which can in certain cases be optimised out as Clang
    // currently does in its lowering. However, for the moment
```

File: mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp
Comment (skatrak) [inline mlir/lib/Target/LLVMIR/Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp:2550]:
```suggestion
    // we do not do so, in part as we currently have substantially less information
    // on the data being mapped at this stage.
```

**Review by skatrak — APPROVED** [review]
Thank you Andrew for working on my previous comments. I have a couple of suggestions to hopefully help simplify things a bit further, but it LGTM. No need to wait for another look by me before merging.

**Comment by agozillon** [comment]
Going to close this current version of the PR stack and open a new one with the changes requested incorporated alongside some newer additions to add support for indexing members at arbitrary depths and some other general fixes, unfortunately lost track of what is in the PR stack and what isn't after working downstream on it for so long, so easier to start fresh to make sure nothing is missed. Incredibly sorry for the bother.

**Pull request closed without merging**
